### DISCUSSION TITLE
Wind turbine curtailment via ROSCO

### DISCUSSION THREAD

**Discussion by lorenzoschena** [NatLabRockies/ROSCO#205] [Announcements]
Hello there, 

I was wondering if it is possible to employ ROSCO for power curtailment purposes. 

A first, naive, way to tune the controller to this end might be to modify the `V_rated` of the turbine model to be used? Is there any drawback that you can see in this approach? 

If yes, can you please point me in the right direction? 

Many thanks,

**Comment by dzalkind** [comment]
Hi,

Modifying `v_rated` in the tuning_yaml will not actually help you de-rate.  `v_rated` is used to separate the range of wind speeds used in the automated tuning process.  It's a common misunderstanding about the input we should probably clarify better.

I would change the `rated_rotor_speed` and/or `rated_power`.  Together, those will determine the rated torque, and the ROSCO tuning process should figure out the proper gains and set points from that.

I hope this helps.

Best, Dan

**Comment by dzalkind** [comment]
I'm going to convert this to a discussion because it's not really an issue with the code, but perhaps others would like to contribute or learn more.

**Comment by lorenzoschena** [comment]
Hi, 
In the context of the discussion above, I was wondering if it was possible to switch between different de-rated controllers during a simulation. From a quick look into the source code, the DLL appears to be loaded statically at the initialization of the simulation, making an "hack" in this direction quite challenging. 

Does ROSCO by any chance support this change, maybe trough ZMQ? If not, do you have any suggestions in this sense? 

Best, 
Lorenzo

**Reply by dzalkind** [reply]
Hi Lorenzo,

Switching between different controller libraries in a single simulation is not possible at this time.  There is a need for this in controller testing environments on deployed turbines, so it's on our radar.  I think the change would need to happen on the OpenFAST side, or at least outside of ROSCO.  I don't think it would be too difficult of a hack, but you will need to be careful about how the controllers initialize.  There may be some updates needed on the ROSCO side to ensure that the library can be loaded in the middle of a simulation.

You may be interested in a new feature we've added to control the power output of a turbine. It's on the develop branch, and you can change the power rating (fraction) with a few different methods, including a ZeroMQ input: https://github.com/NREL/ROSCO/pull/375 

I hope this helps.

Best, Dan

**Reply by lorenzoschena** [reply]
Hi Dan, 

Thanks a lot for the quick response. Indeed, the new functionality of ROSCO seems to be exactly what I was looking for, I am trying it right away. 

Is the WISDEM version changed with respect the main ROSCO requirements? Because I'm getting: 

```
  File "/Users/lorenzoschena/miniconda3/envs/rosco-env-tmp/lib/python3.10/site-packages/rosco/toolbox/ofTools/case_gen/run_FAST.py", line 31, in <module>
    from wisdem.commonse.mpi_tools              import MPI
ModuleNotFoundError: No module named 'wisdem.commonse.mpi_tools'
```

Even though:
```
Name: wisdem
Version: 3.18.0
Summary: Wind-Plant Integrated System Design & Engineering Model
Home-page: https://github.com/WISDEM/WISDEM
Author: 
Author-email: NREL WISDEM Team <systems.engineering@nrel.gov>
License: Apache-2.0
Location: /User/miniconda3/envs/rosco-env-tmp/lib/python3.10/site-packages
Requires: dearpygui, jsonschema, marmot-agents, moorpy, numpy, openmdao, openpyxl, pandas, pydoe3, python-benedict, pyyaml, ruamel.yaml, scipy, simpy, sortedcontainers, statsmodels
Required-by: rosco
```

Best,
Lorenzo

**Reply by lorenzoschena** [reply]
**Follow-up** 

leaning on Openmdao instead of wisdem in `run_FAST` works, e.g. substituted `from wisdem.commonse.mpi_tools              import MPI` with `from openmdao.utils.mpi import MPI` in imports.

**Reply by dzalkind** [reply]
Ah, thank you for the reminder.  That is the recommended fix.  

We should do a hotfix into main.  I opened an issue #399
